### ISSUE TITLE
Appsync resolver to have access to ENV variable

### ISSUE THREAD

**Issue by mattiLeBlanc** [aws-amplify/amplify-category-api#408]

**Comment by ejubber** [comment]
Is there any update on this?

**Comment by zirkelc** [comment]
I'd like to add another issue where support for environment variables is missing:

The IAM authorization for AppSync requires to add all allowed roles or usernames to the `custom-roles.json`: 
```json
{
  "adminRoleNames": ["my-iam-role-dev", "my-iam-role-prod"]
}
```
https://docs.amplify.aws/cli/graphql/authorization-rules/#use-iam-authorization-within-the-appsync-console

These roles will be copied and hardcoded into the generated auth resolvers:
```vtl
#if( $util.authType() == "IAM Authorization" )
  #set( $adminRoles = ["my-iam-role-dev", "my-iam-role-prod"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
  #if( ($ctx.identity.userArn == $ctx.stash.authRole) || ($ctx.identity.cognitoIdentityPoolId == "eu-west-1:..." && $ctx.identity.cognitoIdentityAuthType == "authenticated") )
    #set( $isAuthorized = true )
  #end
#end
```

The IAM roles contain the environment variable `dev` or `prod` and we have currently no pssobility to replace this value with the correct Amplify env. It would be good to have support for the `${env}` syntax that is already being supported for function resolver: https://docs.amplify.aws/cli/graphql/custom-business-logic/#reference-amplify-environment-name

The alternative presented in this issue with a pipeline resolver to add the env to `ctx.stash.env` and then referencing it in the IAM role as `${ctx.stash.env}` might work, but requires to overwrite every single resolver.

**Comment by nkhatiwada7** [comment]
Any updates on this? This is serious limitation. The whole point of amplify is to simplify app-development. I have a custom resolver for `TransactWriteItems` that I'll have to maintain manually until there's a fix. This issue opened over 3 years ago. Please provide a remedy 😞

**Comment by gxxcastillo** [comment]
Not sure if this helps anyone but I needed to know the full table names in the deployed environment and solved it by using override.ts to create a map of all model names to table names and inserting that into all of my resolvers.  It's not elegant and far from optimal but it did unblock me from having to manually update every resolver every time I pushed up new changes.

The way this looked was something like:
1. iterate over every resources.model and creating a map of modelName -> model.modelDBTable.tableName
2. create a template string that puts all table names into the context, something like: `$util.qr($ctx.stash.put("tableNames", ${JSON.stringify(tableNameMap)}))`
3. iterate over the request mapping template for the resolvers, split them by new lines into an array, insert the template string from `2` into that array, and then join the array back into a string.
4. I then assigned that new mapping template to my resolver. `models[modelName].resolvers[resolverName].requestMappingTemplate = newRequestMappingTemplate;`

The end result is i could do something like `"table": $ctx.stash.tableNames.{tableName}` for all my tables in my `TransactWriteItems` operations.

**Comment by chstrong** [comment]
I would also much appreciate the feature to pass environment variables into VTL, just as with Lambda. At the moment I'm stuck. I'm using CDK for the build. The only possibility I see is using Lambda, which is slower.

**Comment by HeskethGD** [comment]
In my case I'm building an AppSync JS Function connected to a Http data source which in this case is a Step Function. We're using CDK and we have Sandbox, Dev, and Prod accounts. We need to pass a different Step Function ARN to the JS code for each one and currently we have to inline the code as a string to inject a different variable. It looks messy, it's error prone and not easily tested. It would be good to be able to pass env variables to the JS function like we do with Lambda functions.

**Comment by joekendal** [comment]
Another use case is updating an OpenSearch index (create new index -> reindex documents) and moving traffic to this new index without downtime. Without the ability to use environment variables, we need to redeploy

**Comment by micchickenburger** [comment]
Another use case is when using an HTTP datasource, like for publishing SNS messages, in which case the AppSync resolver or function needs the SNS Topic ARN.

**Comment by SchollSimon** [comment]
It's really a huge problem for our company, this is urgently needed to use AWS AppSync. Almost not usable in production anymore.

**Comment by mobob** [comment]
+1. Inline code option seems about the only option, and its rather gross. env vars or some kind of build time config please!

**Comment by jeffshep** [comment]
Adding a Terraform workaround for those who don't want to go down the pipeline resolver route.

Defining a UNIT JS resolver in Terraform uses the `code` argument to pass the filename that contains the resolver logic. 
Using a `templatefile` function, instead of the documented `file` function, allows for template syntax and string substitution.

So the request function can look like:
`    operation: "BatchGetItem",
    tables: {
      "${table_name}": {
        keys: ctx.args.id.map((id) => util.dynamodb.toMapValues({ id})),
        consistentRead: true,
      },
    },
`

And the Terraform resource can define:
`code = templatefile("code-directory", {table_name = "table-name})`

**Comment by onlybakam** [comment]
AppSync just released support for Environment Variables: https://docs.aws.amazon.com/appsync/latest/devguide/environmental-variables.html

The feature is live and the Cloudformation docs will be updates shortly.

**Comment by pedroprieto** [comment]
So, how do you use it in a `resource.ts` file in Amplify data? How do you pass the table name to the javascript custom resolver?

File `resource.ts`:
```js
  Item: a
    .model({
      name: a.string(),
    })
    .authorization((allow) => [allow.guest(), allow.authenticated()]),

  itemList: a
    .query()
    .arguments({
      items: a.string().array(),
      /* should I pass an extra argument in here? How do I get the table name? */
    })
    .returns(a.ref("Item").array())
    .handler(
      a.handler.custom({
        dataSource: a.ref("Item"),
        entry: "./item-list.js",
      })
    )
    .authorization((allow) => [allow.authenticated()]),
```

File `item-list.js`:

```js
export function request(ctx) {
  return {
    operation: "BatchGetItem",
    tables: {
      TABLE_NAME: {  /* How do I get the right name in here? */
        keys: ctx.args.items.map((id) => util.dynamodb.toMapValues({ id })),
        consistentRead: true,
      },
    },
  };
}

export function response(ctx) {
  if (ctx.error) {
    util.error(ctx.error.message, ctx.error.type);
  }
  return ctx.result.data.TABLE_NAME;
}
```

**Comment by godrimcom** [comment]
this is so frustrating, why is so complicated to do a BatchGeleteItem using amplify gen 2? How do I get the table name if I already set the dataSource ? please someone help

**Comment by mattiLeBlanc** [comment]
Is this is resolver for only one table? If so, why don't you hardcode it in there? If the name differs per environment you can do an if else to set the table name.

What I am doing, is usinng CDK and when create my resolvers I inject my tables names at the the top of the resolver file so they are available as variables in the resolver on runtime and I can reference them.
Not sure if you use pregenerated amplify code it is as easy?

**Comment by pedroprieto** [comment]
Yes, but when you don't use CDK, there is no way to know the table name when using Amplify...

**Comment by chelorope** [comment]
A simple workaround for Amplify Gen 2 users:

File `backend.ts`:
```
const backend = defineBackend({
  auth,
  data,
  storage,
});

backend.data.resources.cfnResources.cfnGraphqlApi.environmentVariables = {
  API_ID: backend.data.apiId,
};
```

Then you can use `ctx.env.API_ID` inside your resolver function:
```
export function request(ctx) {
  return {
    operation: "BatchPutItem",
    tables: {
      [`YourTableName-${ctx.env.API_ID}-NONE`]:  [item1, item2, ....]
    },
  };
}
```

**Comment by s-a-sureshkumar** [comment]
> A simple workaround for Amplify Gen 2 users:
> 
> File `backend.ts`:
> 
> ```
> const backend = defineBackend({
>   auth,
>   data,
>   storage,
> });
> 
> backend.data.resources.cfnResources.cfnGraphqlApi.addPropertyOverride(
>   "EnvironmentVariables",
>   {
>     API_ID: backend.data.apiId
>   }
> );
> ```
> 
> Then you can use `ctx.env.API_ID` inside your resolver function:
> 
> ```
> export function request(ctx) {
>   return {
>     operation: "BatchPutItem",
>     tables: {
>       [`YourTableName-${ctx.env.API_ID}-NONE`]:  [item1, item2, ....]
>     },
>   };
> }
> ```

It will not work and will throw a 'Circular dependency' error. @chelorope

**Comment by ggbouz** [comment]
> AppSync just released support for Environment Variables: https://docs.aws.amazon.com/appsync/latest/devguide/environmental-variables.html
> 
> The feature is live and the Cloudformation docs will be updates shortly.

The correct answer above

## I'll leave some examples for CDK  typescript users using AppsyncJs. + Dynamodb BatchGetItem

### In your api constructor

When constructing your GraphQL API in CDK, you can set environment variables by using the environmentVariables property:

(you could replace the string with a process.env.TABLENAME This is already a normal JS environment)

```javascript
new GraphqlApi('apiName', {
            //others config for your API
            environmentVariables: {
                "EXAMPLE_TABLE_NAME" :"TableNameValue"
            }
        })
```

### In your AppsyncJs resolver use ctx.env

Within your resolver, you can access the environment variables using ctx.env. Here's an example where the EXAMPLE_TABLE_NAME environment variable is used to interact with a DynamoDB table:

Appsync js has a limited javascript context where environment variables can only be accessed using ctx.env

```javascript
export function request(ctx) {

    return {
        operation: 'BatchGetItem',
        tables: {
            ctx.env.EXAMPLE_TABLE_NAME : {
                keys: util.dynamodb.toMapValues({ pk: 'test', sk: 'test' }),
                consistentRead: true
            },
        },
    }
}

```
